### PR TITLE
Set default dtype of transformed feature to float32

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/transform.py [hunk 1]
```diff
@@ -577,7 +577,11 @@ def parse_feat_ops(confs):
         assert 'feature_col' in feat, \
                 "'feature_col' must be defined in a feature field."
         feat_name = feat['feature_name'] if 'feature_name' in feat else feat['feature_col']
-        out_dtype = _get_output_dtype(feat['out_dtype']) if 'out_dtype' in feat else None
+
+        # By default the out_dtype is set to float32
+        # FeatTransform sub-classes can ignore out_dtype
+        # or redefine it in its init function.
+        out_dtype = _get_output_dtype(feat['out_dtype']) if 'out_dtype' in feat else np.float32
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:584]: what about one-hot encoding? we decide to use int8 to store the data.

File: python/graphstorm/gconstruct/utils.py [hunk 2]
```diff
@@ -304,7 +304,13 @@ def _merge_arrs(arrs, tensor_path):
     """
     assert isinstance(arrs, list)
     shape = _get_tot_shape(arrs)
-    dtype = arrs[0].dtype
+    # To get the output dtype by accessing the
+    # first element of the arrays (numpy array or HDFArray)
+    # Note: We use arrs[0][0] instead of arrs[0] because
+    #       arrs[0][0] is a transformed data with out_dtype
+    #       while arrs[0] can be a HDFArray and has not
+    #       been cast to out_dtype.
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/utils.py:312]: does it mean that HDFArray has different dtype from the data type in array?

File: python/graphstorm/gconstruct/transform.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:584]: The One-hot encoding transformer can ignore this out_dtype and has its own strategy.

Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:584]: The conversion is done inside each transformation.

File: python/graphstorm/gconstruct/utils.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:312]: HDFArray's dtype it is input dtype.

File: python/graphstorm/gconstruct/transform.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:584]: CategoricalTransform just ignore out_dtype.

File: python/graphstorm/gconstruct/transform.py [hunk 3]
```diff
@@ -372,7 +431,7 @@ def after_merge_transform(self, feats):
         feats = np.clip(feats, -1 + self._epsilon, 1 - self._epsilon)
         feats = erfinv(feats)
 
-        feats = _feat_astype(feats, self._out_dtype)
+        self.as_out_dtype(feats)
```
Comment (zheng-da) [inline python/graphstorm/gconstruct/transform.py:434]: do you need to call it here?

File: python/graphstorm/gconstruct/utils.py (hunk 2 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/utils.py:312]: i don't understand. what is input dtype?

Comment (classicsong) [reply python/graphstorm/gconstruct/utils.py:312]: dtype of HDFArray._arr

File: python/graphstorm/gconstruct/transform.py (hunk 3 above)
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:445]: For RankGaussTransform, yes. 
Its feats is transformed in after_merge_transform(), which is after __call__() call.

**Review by zheng-da — APPROVED** [review]

**Pull request merged**